File: leetcode/src/lc19.rs
struct Solution {}

#[cfg(test)]
mod tests {
	use super::*;

	#[test]
	fn test() {
	}
}

use crate::leetcode::ListNode;

impl Solution {
	pub fn remove_nth_from_end(head: Option<Box<ListNode>>, n: i32) -> Option<Box<ListNode>> {
		let mut nil = Some(Box::new(ListNode::new(-1)));
		nil.as_mut().unwrap().next = head;
		Solution::dfs(&mut nil, n);
		nil.unwrap().next
	}

	fn dfs(node: &mut Option<Box<ListNode>>, target: i32) -> i32 {
		if let Some(node) = node {
			let ret = Solution::dfs(&mut node.next, target);
			if ret == target {
				let nxt = &node.next.as_ref().unwrap().next;
				node.next = nxt.clone()
			}
			ret + 1
		} else {
			0
		}
	}
}

File: src/leetcode/lc19.py
from leetcode import ListNode

class Solution:
	def removeNthFromEnd(self, head: ListNode, n: int) -> ListNode:
		nil = ListNode(0)
		nil.next = head

		def dfs(prev, cur):
			if cur is None:
				return 1
			idx = dfs(cur, cur.next)
			print(cur.val, idx)
			if idx == n:
				prev.next = cur.next
			return idx + 1

		dfs(nil, head)
		return nil.next
